xrootd/client/protocol.go

```diff
@@ -0,0 +1,93 @@
+// Copyright 2018 The go-hep Authors.  All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package client // import "go-hep.org/x/hep/xrootd/client"
+
+import (
+	"context"
+
+	xrdproto "go-hep.org/x/hep/xrootd/protocol"
+	"go-hep.org/x/hep/xrootd/protocol/protocol"
+)
+
+// ProtocolInfo is a response for the `Protocol` request. See details in the xrootd protocol specification.
+type ProtocolInfo struct {
+	BinaryProtocolVersion int32
+	ServerType            xrdproto.ServerType
+	IsManager             bool
+	IsServer              bool
+	IsMeta                bool
+	IsProxy               bool
+	IsSupervisor          bool
+	SecurityVersion       byte
+	ForceSecurity         bool
+	SecurityLevel         protocol.SecurityLevel
+	SecurityOverrides     []protocol.SecurityOverride
+}
+
+// Protocol obtains the protocol version number, type of the server and security information, such as:
+// the security version, the security options, the security level, and the list of alterations
+// needed to the specified predefined security level.
+func (client *Client) Protocol(ctx context.Context) (ProtocolInfo, error) {
```

Comment:
I am wondering whether `Protocol` should be a method on `xrootd/client.Client` or a function of `xrootd/protocol/protocol`...

if we keep `Protocol` a method of `client.Client`, then I suppose we'll have to have every single request/response as a method of `client.Client` as well (_ie_ 32 methods)

I am thinking out loud here (so it may just be a completely bunker idea), but it seems we could just get away with 1 or 2 methods in `client.Client` and keep the rest in each of the `xyz` protocol packages.
it *seems* to me this would introduce less coupling between `client.Client` and each request/response.

for example:

- `xrootd/protocol/protocol`:
```go
type Request { ... }
func NewRequest(...) Request { ... }

type Response { ... } // essentially client.ProtocolInfo

func Send(ctx context.Context, cli *xrdclient.Client, req Request) (Response, error) {
    raw, err := cli.Call(ctx, req.ID(), req.Marshal(cli.Version()))
    // ...
    var resp Response
    err = xrdproto.Unmarshal(&resp)
    return resp, err
}
```

I thought that was what you were aiming at when you put each request-type into its own package.

that said, I am not completely happy with the `Send` function as it is.
Perhaps, `Send` should actually be a method of `xrdclient.Client`, but it would take a (yet to be defined) `xrdproto.Request` interface as argument:

```go
func (cli *Client) Send(ctx context.Context, resp Response, req Request) error {
    raw, err := cli.marshal(req)
    recv, err := cli.call(ctx, req.ID(), raw)
    err = cli.unmarshal(raw, resp)
    return err
}
```
and used like so:
```go
var resp protocol.Response
err := cli.Send(ctx, &resp, protocol.NewRequest(...))
```

from a user perspective, it would look better like so:
```go
resp, err := cli.Send(ctx, protocol.NewRequest(...))
```
but one would have to type-assert to get the `protocol.Response` type...

I am still on the fence.
happy to get your input :)

Reply:
It sounds reasonable to me. 
In fact, most of these `Client.*` request methods will be pretty straightforward - create request, send it, parse and return response.

However, from the user perspective it is:
```go
dirs, err := client.Dirlist(ctx, "\tmp")
```
vs
```go
var resp dirlist.Response
err := client.Send(ctx, dirlist.NewRequest("\tmp"))
dirs := resp.Dirs
// (or without predeclaration but with type-assertion)
```
So we have a lot more boilerplate code from user perspective IMHO.
Also, it's much easier to type "client." and get a list of code completions than be like "What was the name of that request again?".

Actually, I think, we can implement both ways:
- After implementing a few more requests and the marshalling code generator, implement routine like this:
```go
 raw, err := client.marshal(req)
 recv, err := client.call(...)
 err = client.unmarshal(...)
 return err
```
- Make small helper methods of `Client`:
```go
 func (client *Client) Dirlist(ctx context.Context, path string) ([]string, err) {
     var resp dirlist.Response
     err := client.Send(ctx, &resp, dirlist.NewRequest(path))
     if err != nil {
         return nil, err
     }
     return resp.Dirs, nil
 }
```

What do you think about it?